docker/docker_image_src.go

```diff
@@ -28,17 +29,91 @@ type dockerImageSource struct {
 }
 
-// newImageSource creates a new ImageSource for the specified image reference.
-// The caller must call .Close() on the returned ImageSource.
-func newImageSource(sys *types.SystemContext, ref dockerReference) (*dockerImageSource, error) {
-	c, err := newDockerClientFromRef(sys, ref, false, "pull")
+// newImageSource creates a new `ImageSource` for the specified image reference
+// `ref`.
+//
+// The following steps will be executed in order during the instance creation:
+//
+// 1. Lookup the registry within the configured location in
+//    `sys.SystemRegistriesConfPath`. If there is no configured registry available,
+//	  we create one by our own based on the provided docker reference `ref`.
+//
+// 2. References which contain a configured prefix will be automatically rewritten
+// 	  to the correct target reference. For example, if the configured
+// 	  `prefix = "example.com/foo"`, `url = "example.com"` and the image will be
+// 	  pulled from the ref `example.com/foo/image`, then the resulting pull will be
+// 	  effectively point to `example.com/image`.
+//
+// 3. If the rewritten reference succeeds, it will be used as the `dockerRef`
+// 	  in the client. If not, we fall back to using the initial `ref`.
+//
+// 4. Each mirror will be used (in the configured order) to test the
+// 	  availability of the image manifest on the remote location. For example,
+// 	  if the manifest is not reachable due to connectivity issues, then the next
+// 	  mirror will be tested instead. If no mirror is configured or contains the
+// 	  target manifest, then the initial `ref` will be tested as fallback. The
+// 	  creation of the new `dockerImageSource` only succeeds if a remote
+// 	  location with the available manifest was found.
+//
+// A cleanup call to `.Close()` is needed if the caller is done using the returned
+// `ImageSource`.
+func newImageSource(ctx context.Context, sys *types.SystemContext, ref dockerReference) (*dockerImageSource, error) {
+	registry, err := sysregistriesv2.FindRegistry(sys, ref.ref.Name())
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err, "error loading registries configuration")
 	}
-	return &dockerImageSource{
-		ref: ref,
-		c:   c,
-	}, nil
+
+	if registry == nil {
+		// No registry was found by the provided reference, so we create a
+		// fallback registry by hand to make the client creation below work as
+		// intended.
+		registry = &sysregistriesv2.Registry{
+			Endpoint: sysregistriesv2.Endpoint{
+				URL: ref.ref.String(),
+			},
+		}
+	}
+
+	// Found the registry within the sysregistriesv2 configuration. Now we test
+	// all endpoints for the manifest availability. If an working image source
+	// was found, it will be used for all future pull actions.
+	var (
+		imageSource     *dockerImageSource
+		manifestLoadErr error
+	)
+	for index, endpoint := range append(registry.Mirrors, registry.Endpoint, registry.Endpoint) {
```

Comment:
I've did this because of the fallback testing. So we still rewrite the original ref, but if that will result in a wrong remote location, the untouched original ref will be used:
```go
		if index <= len(registry.Mirrors) {
			newRef, err := endpoint.RewriteReference(ref.ref, registry.Prefix)
			if err != nil {
				return nil, errors.Wrapf(err, "error in newImageSource")
			}
			dockerRef = dockerReference{newRef}
		}
```

Reply:
Let's drop that logic. Sorry for requesting that change earlier; I did not think this through.

The reason why I changed my mind is considering the two cases of a prefix. Either it is unspecified which means that `prefix == url` which means that we already have the endpoint as a "fallback". Or we specified a prefix which implies that we want to pull from somewhere else. Falling back to `url` conflicts with that. If a user wishes to also attempt to pull from `url`, it can be added as a mirror.

Reply:
Sure, I adapt the implementation as suggested.